homeassistant/components/deconz/fan.py

```diff
@@ -10,22 +13,27 @@
 )
 from homeassistant.core import callback
 from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.util.percentage import (
+    ordered_list_item_to_percentage,
+    percentage_to_ordered_list_item,
+)
 
 from .const import FANS, NEW_LIGHT
 from .deconz_device import DeconzDevice
 from .gateway import get_gateway_from_config_entry
 
-SPEEDS = {SPEED_OFF: 0, SPEED_LOW: 1, SPEED_MEDIUM: 2, SPEED_HIGH: 4}
-SUPPORTED_ON_SPEEDS = {1: SPEED_LOW, 2: SPEED_MEDIUM, 4: SPEED_HIGH}
+ORDERED_NAMED_FAN_SPEEDS = [1, 2, 3, 4]
+
+LEGACY_SPEEDS = {SPEED_OFF: 0, SPEED_LOW: 1, SPEED_MEDIUM: 2, SPEED_HIGH: 4}
```

Comment:
If there isn't a legacy speed to map to `3` everything will be off by one which is likely the reason the test is mapping the speed up 1

Reply:
Yes it wasnt possible to map 3 before so that was ignored and replaced by medium

Reply:
To avoid the off by one and getting the wrong speed there are two options I can think of that would make the translation layer work as expected:

1. Add something to the speed this that would map to speed 3.  This would technically violate the old entity model, but I'm not sure how much of an issue that is since we are only doing it for backwards compatibility and its going away anyways.

`LEGACY_SPEEDS = {SPEED_OFF: 0, SPEED_LOW: 1, SPEED_MEDIUM: 2, SPEED_MED_HIGH: 3, SPEED_HIGH: 4}`

2. (Preferred).  Override the bound methods of `self.percentage_to_speed` and `self.speed_to_percentage`. This would allow you to map them manually to ensure you always get the behavior you want.

Reply:
I think in this case it would be Ok to override the compat method to get the legacy support working (for a time)

Reply:
Fixed